drm-howto/modeset-atomic.c

```diff
@@ -0,0 +1,1073 @@
+ * be updating other planes in this same atomic commit.
+ */
+
+static void modeset_draw_dev(int fd, struct modeset_dev *dev)
+{
+	struct modeset_buf *buf;
+	unsigned int j, k, off;
+	int ret;
+
+	dev->r = next_color(&dev->r_up, dev->r, 5);
+	dev->g = next_color(&dev->g_up, dev->g, 5);
+	dev->b = next_color(&dev->b_up, dev->b, 5);
+
+	buf = &dev->bufs[dev->front_buf ^ 1];
+	for (j = 0; j < buf->height; ++j) {
+		for (k = 0; k < buf->width; ++k) {
+			off = buf->stride * j + k * 4;
+			*(uint32_t*)&buf->map[off] =
+				     (dev->r << 16) | (dev->g << 8) | dev->b;
+		}
+	}
+
+	ret = modeset_atomic_commit(fd, buf, dev);
+	if (ret) {
+		fprintf(stderr, "atomic commit failed, %d\n", errno);
+	} else {
+		dev->front_buf ^= 1;
+		dev->pflip_pending = true;
+	}
+}
+
+/*
+ * modeset_cleanup() stays the same.
+ */
+
+static void modeset_cleanup(int fd)
+{
+	struct modeset_dev *iter;
+	drmEventContext ev;
+	int ret;
+
+	/* init variables */
+	memset(&ev, 0, sizeof(ev));
+	ev.version = DRM_EVENT_CONTEXT_VERSION;
+	ev.page_flip_handler = modeset_page_flip_event;
+
+	while (modeset_list) {
+		/* remove from global list */
+		iter = modeset_list;
+		modeset_list = iter->next;
+
+		/* if a page-flip is pending, wait for it to complete */
+		iter->cleanup = true;
+		fprintf(stderr, "wait for pending page-flip to complete...\n");
+		while (iter->pflip_pending) {
+			ret = drmHandleEvent(fd, &ev);
+			if (ret)
+				break;
+		}
+
+		/* restore saved CRTC configuration */
+		if (!iter->pflip_pending)
+			drmModeSetCrtc(fd,
+				       iter->saved_crtc->crtc_id,
+				       iter->saved_crtc->buffer_id,
+				       iter->saved_crtc->x,
+				       iter->saved_crtc->y,
+				       &iter->conn_id,
+				       1,
+				       &iter->saved_crtc->mode);
+		drmModeFreeCrtc(iter->saved_crtc);
+
+		/* destroy framebuffers */
+		modeset_destroy_fb(fd, &iter->bufs[1]);
+		modeset_destroy_fb(fd, &iter->bufs[0]);
+
+		/* free allocated memory */
+		free(iter);
+	}
+}
+
+/* 
+ * This example shows how to use the KMS atomic API to handle page-flips.
+ * Besides the fact that we've only used the primary plane, you could extend
+ * the example to use overlay and cursor planes. The same approach could be
+ * used to do a page-flip: all the planes are updated in local objects and
+ * then drmModeAtomicCommit() can be called to perform a page-flip.
+ * 
+ * Just like vsync'ed double-buffering, the atomic API does not not solve all
+ * the problems that can happen and you have to figure out the best
+ * implementation for your use case.
+ *
+ * If you want more code, I can recommend reading the source-code of:
+ *  - plymouth (which uses dumb-buffers like this example; very easy to understand)
+ *  - kmscon (which uses libuterm to do this)
+ *  - wayland (very sophisticated DRM renderer; hard to understand fully as it
+ *             uses more complicated techniques like DRM planes)
+ *  - xserver (very hard to understand as it is split across many files/projects)
```

Comment:
Drop xserver from here.

People would probably go looking at the modesetting Xorg video DDX which lives in the xserver repository, and but it does not do atomic KMS. Well, it tried, and failed horribly enough that the kernel is actually blacklisting X for atomic KMS: https://lists.freedesktop.org/archives/dri-devel/2019-September/234193.html

I don't know if hardware-specific video DDXes did atomic right, but they would be hit by the blacklist too.

Reply:
Done.